planner/optimize.go

```diff
@@ -162,6 +162,12 @@ func Optimize(ctx context.Context, sctx sessionctx.Context, node ast.Node, is in
 		}
 	}
 
+	// Override the resource group if necessary
+	// TODO: we didn't check the existence of the hinted resource group now to save the cost per query
+	if originStmtHints.HasResourceGroup {
+		sessVars.ResourceGroupName = originStmtHints.ResourceGroup
```

Comment:
What about use put the resource group hint into StatementContext, so we don't need to change and restore the session variable.

Reply:
As we talked offline, your proposed change doesn't simplify the work.